Generate an error on unknown Dockerfile instruction

duglin opened the pull request:
Instead of just printing a warning and going on, this will generate
an error and stop processing.

This used to be part of #10561 but I decided it might need its own
independent discussion/PR as to not derail #10561.

Signed-off-by: Doug Davis dug@us.ibm.com

Commit:
Generate an error on unknown Dockerfile instruction

Instead of just printing a warning and going on, this will generate
an error and stop processing.

This used to be part of #10561 but I decided it might need its own
independent discussion/PR as to not derail #10561.

Signed-off-by: Doug Davis <dug@us.ibm.com>

jfrazelle commented:
Aren't comments considered unknown

On Thursday, February 5, 2015, Doug Davis <notifications@github.com> wrote:

> Instead of just printing a warning and going on, this will generate
> an error and stop processing.
>
> This used to be part of #10561
> <https://github.com/docker/docker/pull/10561> but I decided it might need
> its own
> independent discussion/PR as to not derail #10561
> <https://github.com/docker/docker/pull/10561>.
>
> Signed-off-by: Doug Davis dug@us.ibm.com
> <javascript:_e(%7B%7D,'cvml','dug@us.ibm.com');>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/docker/docker/pull/10591
> Commit Summary
>
>    - Generate an error on unknown Dockerfile instruction
>
> File Changes
>
>    - *M* builder/evaluator.go
>    <https://github.com/docker/docker/pull/10591/files#diff-0> (4)
>    - *M* integration-cli/docker_cli_build_test.go
>    <https://github.com/docker/docker/pull/10591/files#diff-1> (4)
>
> Patch Links:
>
>    - https://github.com/docker/docker/pull/10591.patch
>    - https://github.com/docker/docker/pull/10591.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10591>.
>

jfrazelle commented:
What about MAINTAINERS or DOCKER-VERSION idk if people use other things too

On Thursday, February 5, 2015, Jessica Frazelle <jess@docker.com> wrote:

> Aren't comments considered unknown
>
> On Thursday, February 5, 2015, Doug Davis <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> Instead of just printing a warning and going on, this will generate
>> an error and stop processing.
>>
>> This used to be part of #10561
>> <https://github.com/docker/docker/pull/10561> but I decided it might
>> need its own
>> independent discussion/PR as to not derail #10561
>> <https://github.com/docker/docker/pull/10561>.
>>
>> Signed-off-by: Doug Davis dug@us.ibm.com
>> ------------------------------
>> You can view, comment on, or merge this pull request online at:
>>
>>   https://github.com/docker/docker/pull/10591
>> Commit Summary
>>
>>    - Generate an error on unknown Dockerfile instruction
>>
>> File Changes
>>
>>    - *M* builder/evaluator.go
>>    <https://github.com/docker/docker/pull/10591/files#diff-0> (4)
>>    - *M* integration-cli/docker_cli_build_test.go
>>    <https://github.com/docker/docker/pull/10591/files#diff-1> (4)
>>
>> Patch Links:
>>
>>    - https://github.com/docker/docker/pull/10591.patch
>>    - https://github.com/docker/docker/pull/10591.diff
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/docker/pull/10591>.
>>
>

cpuguy83 commented:
I'm positive we've discussed this one before and decided to not do it b/c it will definitely break people.

LK4D4 commented:
@duglin Yeah, this is awful, but we need keep it :(

duglin commented:
Comments are not considered commands so they're still ignored.

If it breaks people then that's good!  IMO, if someone added something to a Dockerfile they should want to know if the daemon/builder is going to ignore it - if it can be ignored then why add it in the first place?  I consider it a bug to NOT flag bad/unknown commands.

However, if we really do want this behavior I would prefer to make it available via a "docker build" option and NOT set as the default.

But this is why I pulled it out of #10561, I knew it might cause some interesting discussions. :-)

duglin commented:
I'd be really interested in hearing from people who use/like this current behavior - maybe that'll help me understand why we shouldn't consider this a bug.

LK4D4 commented:
@duglin Yes, this is a bug, but now people have thousands of invalid commands in their Dockerfiles in production. And broke productions is not what we want very much in favor of fixing minor bug(even if it is pretty annoying).

duglin commented:
Do we know for sure that people have broken Dockerfiles?

bacongobbler commented:
Hmm... Maybe an option flag for `docker build` would suffice? `sphinx-doc` has a `-W` flag which turns warnings into errors.

duglin commented:
Right, if it turns out we can't do the right thing  :-)  then I'd like to add a flag to turn these into errors so that people can start the process of fixing their Dockerfiles - and, at least via docs, mark the current behavior as "deprecated".

thaJeztah commented:
I'm in favor on having an way to enable this by default somehow (env var? daemon flag?)

I've been bitten a few times by docker "silently" skipping lines because I had a whitespace before the command that I didn't notice. (to be honest, docker *does* output a warning, but it's easy lost in the output of the build)

thaJeztah commented:
Oh, a Docker "validate" (or `docker build --validate`) command would be a nice companion; being able to check the Dockerfile *before* attempting to build it. 

I realize that's a separate request :)

duglin commented:
@bacongobbler comments would not generate errors

duglin commented:
Can someone provide a URL to a real Dockerfile that would be broken by this, but works today?

thaJeztah commented:
> Can someone provide a URL to a real Dockerfile that would be broken by this, but works today?

Dockers own Dockerfile(s) before 1.4.0; https://github.com/docker/docker/blob/v1.3.3/Dockerfile

(See this PR: https://github.com/docker/docker/pull/8673/files)

duglin commented:
@thaJeztah well, perhaps I should have been more explicit... I'd like to see one that's actually used today :-)  
Even so, I'm still hoping someone can give me a real life example where they actually want today's behavior.  I agree this is a breaking change but if no one really cares...

jfrazelle commented:
Well I could see people doing it and not even realizing or doing it for
meta data for their own purposes when tho it does not effect the build

On Thu, Feb 5, 2015 at 6:41 PM, Doug Davis <notifications@github.com> wrote:

> @thaJeztah <https://github.com/thaJeztah> well, perhaps I should have
> been more explicit... I'd like to see one that's actually used today :-)
>
> Even so, I'm still hoping someone can give me a real life example where
> they actually want today's behavior. I agree this is a breaking change but
> if no one really cares...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10591#issuecomment-73173978>.
>

duglin commented:
For the first group I suspect (after the initial wtf!) they will he happy to fix their Dockerfiles and will be upset they never noticed the warnings  ;-)
For the 2nd group.... I dunno. Sounds like speculation and if they really do want that I would think sticking it into comments would be safer/better/smarter...  but this is all speculation.  Still want to hear from someone who is counting on this (dare I say "wrong") behavior in a real environment.  I'm sure there's someone out there, just want to hear from them.

jfrazelle commented:
true maybe @sirupsen i know you all had some custom builder?

On Thu, Feb 5, 2015 at 6:48 PM, Doug Davis <notifications@github.com> wrote:

> For the first group I suspect (after the initial wtf!) they will he happy
> to fix their Dockerfiles and will be upset they never noticed the warnings
> ;-)
> For the 2nd group.... I dunno. Sounds like speculation and if they really
> do want that I would think sticking it into comments would be
> safer/better/smarter... but this is all speculation. Still want to hear
> from someone who is counting on this (dare I say "wrong") behavior in a
> real environment. I'm sure there's someone out there, just want to hear
> from them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10591#issuecomment-73174474>.
>

Sirupsen commented:
@jfrazelle We don't use Dockerfiles at all, so this won't cause any issues for us. Thanks for thinking of us :)

thaJeztah commented:
> perhaps I should have been more explicit... I'd like to see one that's actually used today :-)

Yes, I got that. I was *partly* serious thought; would it be a problem if people downloaded the Docker 1.3.3 source and not being able to build it (without fixing the Dockerfile)?

(Perhaps *enable by default* and have a flag to skip validation)

duglin commented:
if we can't just flat out remove it, then yes my 2nd choice would be to generate errors by default but have a flag to revert to today's behavior.  3rd choice is flag with default set to the "incorrect" choice  :-)

cpuguy83 commented:
New flag `--strict=true`

thaJeztah commented:
@cpuguy83 If it's *disabled* by default, then preferably `--no-strict[=false]`.

Boolean flags with a default "true" are confusing.

That's just details though :)

tianon commented:
+1 -- I've actually been running my local Docker with a variant of this patch for several months now (since at least `Nov 22`, which is the mtime on the patch file):
```patch
$ cat /etc/portage/patches/app-emulation/docker/unknown-instructions-error-out.patch
diff --git a/builder/evaluator.go b/builder/evaluator.go
index 645038b..086d440 100644
--- a/builder/evaluator.go
+++ b/builder/evaluator.go
@@ -230,7 +230,5 @@ func (b *Builder) dispatch(stepN int, ast *parser.Node) error {
 		return f(b, strs, attrs, original)
 	}
 
-	fmt.Fprintf(b.ErrStream, "# Skipping unknown instruction %s\n", strings.ToUpper(cmd))
-
-	return nil
+	return fmt.Errorf("Unknown instruction %q\n", strings.ToUpper(cmd))
 }
```

duglin commented:
So... to make sure we're on the same page, are we talking about a daemon flag or CLI flag (on `docker build`)?  I'd prefer daemon flag.

@thaJeztah Just curious, why do you think boolean flags with a default of true are confusing?  I actually, tend to think of flags with "no" in their name as confusing because it feels like we're dealing with a double negative when you say:  --no-foo=false    
So, in this case `--strict=false` is hurts my head less than `--no-strict=true`

tianon commented:
because boolean flags can be specified without a value to imply "true", so
it's really "--strict=true" versus "--no-strict"

duglin commented:
@tianon ok - that makes sense

thaJeztah commented:
Yes, what @tianon said, also see https://github.com/docker/docker/issues/9406 (and some issues related to that). Sorry I wasn't clear.

fwiw, I'm +1 on making it a daemon flag.

SvenDowideit commented:
+1, though it might be useful to make this `--strict=off/false` by default, `--strict=warn` and `--strict=error/true`

that way, the Docker Hub could turn on `strict=warn` to help users.

thaJeztah commented:
> +1, though it might be useful to make this --strict=off/false by default, --strict=warn and --strict=error/true

Nice thinking, that would circumvent the "boolean option/naming" issue.

Bit in doubt; it would overlap with the `--log-level` option, so this may need a bit of thinking

tianon commented:
Personally, I'd much rather see us make the breaking change and finally put
this particular mis-feature behind us for good instead of adding YADF (Yet
Another Daemon Flag), especially one that's so attractive that everyone
really ought to be setting it, but IANTM.

duglin commented:
@tianon I would prefer to just fix this "bug" too :-)
@crosbymichael @vieux @unclejack @tiborvass   thoughts?

unclejack commented:
I'm +1 for fixing, but we need to make sure we start letting people know about it so they can update their Dockerfiles. The longer we wait to fix this, the more likely it is for the number of Dockerfiles doing this to grow.

tiborvass commented:
@duglin @unclejack what would be the deprecation plan?

unclejack commented:
@tiborvass I was thinking we could mention it soon after releasing 1.5 and do it in 1.6.

duglin commented:
options:
1) just fix it, per this PR and people fix their Dockerfiles
2) add a daemon flag that allows for the old/bad behavior - IOW, default is to error on bad cmd
3) add a daemon flag that allows for the correct behavior - IOW, default is today's behavior

I would prefer to do 1, but I think 2 is probably the best just to keep people from pulling their hair out immediately.  But in the end, by 1.7 we pull the daemon flag.  i don't like 3 at all - it'll just encourage people to NOT fix things.

thaJeztah commented:
fwiw, my vote goes to option 2 (best of both worlds), with the addition that option 2 should log invalid commands more verbose to encourage people to fix their Dockerfiles.

jfrazelle commented:
i would almost veer towards 1 because then we can never get rid of the flag

thaJeztah commented:
> i would almost veer towards 1 because then we can never get rid of the flag

I share your concern on that point, but *not* providing a flag could result in people holding of updating Docker altogether, which is (I think) even more of a PITA to keep supporting.

tiborvass commented:
@duglin does this bust cache?

duglin commented:
hmm I don't think it would - unless you consider stopping the build upon reading an unknown command as busting the cache.  :-)
But right now unknown commands are just ignored and so they do not touch/update the cache at all.

tiborvass commented:
@duglin so basically, if a dockerfile with this PR fails because of an unknown command, if the user removes or comments out that command from the dockerfile and redoes a build, I assume the cache will get invalidated.

duglin commented:
No, the cache will not be invalidated.  Here's the output they would see:
```
$ docker build --no-cache tmp   # using master code
Sending build context to Docker daemon 95.23 kB
Sending build context to Docker daemon 
Step 0 : FROM ubuntu
 ---> b39b81afc8ca
Step 1 : NADA 
# Skipping unknown instruction NADA
 ---> b39b81afc8ca
Step 2 : RUN echo hi
 ---> Running in 10137fdf44f9
hi
 ---> edf6d04aefc4
Removing intermediate container 10137fdf44f9
Successfully built edf6d04aefc4

$ vi tmp/Dockerfile   # remove NADA cmd and switch to new code in PR

$ docker build tmp
Sending build context to Docker daemon 95.23 kB
Sending build context to Docker daemon 
Step 0 : FROM ubuntu
 ---> b39b81afc8ca
Step 1 : RUN echo hi
 ---> Using cache
 ---> edf6d04aefc4
Successfully built edf6d04aefc4
```
Notice the cache wasn't invalidated.

tiborvass commented:
@duglin perfect, it makes sense since we were just skipping it, thanks!

duglin commented:
I modified this PR so that there's now a daemon --no-strict flag (default is 'false') which mean it'll stop upon unknown command.  All, PTAL.

SvenDowideit commented:
nice - +1 for generating a warning - this needs to be in the release notes, with some search engine fodder so that users can search by pasting the warning or error text and get the right help

v nice :)

duglin commented:
@SvenDowideit I added something to the release notes.  Should I have removed all of the v1.5 "New Features" lines there?
Also, in your comment you mentioned a warning, was that a +1 to what's there already or were you asking for another warning to be added?

SvenDowideit commented:
+1 to the fact you're generating one :)

wrt release notes - idk, if you would just pop it at the bottom in a 1.6.0 section, we should hopefully we won't forget it

duglin commented:
@SvenDowideit I decided to just tag that new feature entry - hopefully that will stand out enough.

duglin commented:
@tiborvass @jfrazelle @cpuguy83   thoughts on this one?
I'll work on the rebase, but we need a decision on the design.

jfrazelle commented:
no flag just "fix the bug"

duglin commented:
@jfrazelle LOL you're so harsh I love it

cpuguy83 commented:
Yeah, let's just fix it, there's no tests checking to see if this stuff gets ignored as far as I can tell anyway.

duglin commented:
ping @crosbymichael your thoughts? I'm ok with no flag - I added it to give people an "out" until they could fix all of their Dockerfiles (that's why its also deprecated).  Just need an official statement of direction.

jfrazelle commented:
@duglin hehe ;)

duglin commented:
@jfrazelle @cpuguy83 would you object to leaving the daemon-side option there just for one release? Default would be to generate an error.  I'm a bit nervous about not giving people an easy "out" if they run into issues.  Hopefully no one will need it and then removing it in the next release won't be an issue.

thaJeztah commented:
@duglin if it will be removed soon, I think it should be marked "deprecated" from the start, so that people don't start relying on it

duglin commented:
@thaJeztah yup - in the doc updates I made I included the "d" word :-)  e.g. https://github.com/docker/docker/pull/10591/files#diff-c13a5c583fae16a859a034cdd06c7c58R426

duglin commented:
ping @jfrazelle

jfrazelle commented:
shall we move into code review

duglin commented:
Yes please 😄

-Doug

Sent from my iPhone

> On Mar 11, 2015, at 2:48 PM, Jessie Frazelle <notifications@github.com> wrote:
> 
> shall we move into code review
> 
> —
> Reply to this email directly or view it on GitHub.
>

tiborvass commented:
@jfrazelle @duglin it's not clear to me what was agreed on in the design discussion.

duglin commented:
Well, I would _really_ prefer what's in the code - meaning provide them "an out" for one release by offering the (deprecated) --no-strict flag.  @jfrazelle you ok with that?  On the positive side, having it doesn't shouldn't do any harm.

tiborvass commented:
@duglin `--no-strict` is not deprecated in your code. Did you want to introduce a deprecated flag? Or do you intend to deprecate it in a later release?

duglin commented:
@tiborvass sorry, I mentioned it as deprecated in the docs but forgot about \# in the flag definition - fixed now.

tiborvass commented:
@docker/maintainers So are we all okay with introducing a new flag that's deprecated?

I'm keeping it design review until it's clear we all agree, sorry :(

jfrazelle commented:
no flag

icecrime commented:
I don't see a better way for such breaking transition.

@jfrazelle What do you suggest? That we ship the breaking change and that's it?

jfrazelle commented:
It's a "bug" that should be fixed w no flag

On Wednesday, March 11, 2015, Arnaud Porterie <notifications@github.com>
wrote:

> I don't see a better way for such breaking transition.
>
> @jfrazelle <https://github.com/jfrazelle> What do you suggest? That we
> ship the breaking change and that's it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10591#issuecomment-78422861>.
>

thaJeztah commented:
>It's a "bug" that should be fixed w no flag

But breaking the Dockefile "API" in a .X release without giving people time to adopt, sounds troublesome to me. It's not a bug that was *introduced* after the 1.0 release, it's been there all the time.

tianon commented:
I agree with @jfrazelle -- the flag seems kind of pointless, especially
since it's already been a warning for so many versions.  As I mentioned
earlier in this thread, I've been running for months with a local patch
similar to this one, and I think the benefits far, far outweigh the
downsides here.

jamtur01 commented:
+1 to @tianon and @jfrazelle

thaJeztah commented:
> it's already been a warning for so many versions.

The "warning" is `# Skipping unknown instruction HELLO`, which isn't very explicit that this will *break* in future.

The majority here seems to be in favour of *not* adding a flag, though. I'm very much `+1` for this PR, so it's fine with me (fwiw) to drop the flag. Now let's hope it won't cause too many problems.

tiborvass commented:
@thaJeztah: @duglin is testing 1800 dockerfiles across github and bitbucket. Impatient to know the results!

thaJeztah commented:
> is testing 1800 dockerfiles across github and bitbucket. Impatient to know the results!

WOW, so much :heart:

Hope if works and then..  🚩🔫 👍

(May make a great blog-post as well if some nice stats/graphs can be obtained from that)

duglin commented:
https://gist.github.com/duglin/cc2dedbabd6a46f8bdbb

duglin commented:
Most appear to be due to lack of \\ at the end of the previous line.

duglin commented:
Removed --no-strict flag.

tiborvass commented:
So @duglin's test concludes that, there was only Dockerfile out of 1800, who used the feature with a `VERSION` instruction. The other errors should indeed be errors and we should not "skip" them.

So that gist, shows indeed that this default behavior is desired, it would catch errors earlier, and it outweighs the convenience of skipping instructions like VERSION.

The workaround for those Dockerfiles is just to comment out those instructions.

So +1 to removing the flag, I'm moving to code review.

tiborvass commented:
LGTM

thaJeztah commented:
Thanks for approaching this so thoroughly! LGTM (not a maintainer)

icecrime commented:
Nice work! LGTM

duglin commented:
ping @moxiegirl @SvenDowideit @fredlf  for a quick doc review.

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -427,7 +427,6 @@ Docker supports softlinks for the Docker data directory
     export DOCKER_TMPDIR=/mnt/disk2/tmp
     /usr/local/bin/docker -d -D -g /var/lib/docker -H unix:// > /var/lib/boot2docker/docker.log 2>&1
 
-
```
What's this?

duglin replied:
Hmm, I think when I removed the text that I added I removed the extra line.  Added it back in.

jamtur01 commented on docs/sources/release-notes.md:
```diff
@@ -15,6 +15,9 @@ For a complete list of patches, fixes, and other improvements, see the
 *New Features*
 
+* [1.6] The Docker daemon will no longer ignore unknown commands
+  while processing a Dockerfile. Instead it will generate an error and halt
```
\`Dockerfile\`

duglin replied:
fixed

jamtur01 commented:
Minor comments.

duglin commented:
@jamtur01 updated per your comments

jamtur01 commented:
LGTM albeit the tests are now failing. :)

duglin commented:
@jamtur01 yea but that's just windows and not due to this change.

moxiegirl commented:
LGTM
